### PR TITLE
Try to make symbol table access thread safe.

### PR DESCRIPTION
**Pull request by yuyichao**
Tested on top of #14190 and #14361 (it easily segfaults for other reasons otherwise...) on x64.

Also removed the GC write barrier on the symbols since we don't GC them anymore.

I'm fairly sure this is safe on other architechtures that we support while avoiding locking the read path but it would be better if someone with more experience could have a look.

@vtjnash @ArchRobison

### CODE REVIEW

File: src/alloc.c [hunk 1]
```diff
@@ -407,40 +410,56 @@ static jl_sym_t *mk_symbol(const char *str, size_t len)
...
-    jl_sym_t *parent;
-    pnode = symtab_lookup(&symtab, str, len, &parent);
-    if (*pnode == NULL) {
-        *pnode = mk_symbol(str, len);
-        if (parent != NULL)
-            jl_gc_wb(parent, *pnode);
+    jl_sym_t *volatile *slot;
+    jl_sym_t *node = symtab_lookup(&symtab, str, len, &slot);
+    if (node == NULL) {
+        JL_LOCK(symbol_table); // Might GC
```
Comment (yuyichao) [inline src/alloc.c]: Note that this `// Might GC` comment only applies after #14190.

**Comment by stevengj** [comment]
Shouldn't we just use some mature concurrent tree or hash-table library here?  e.g. [nbds](https://code.google.com/p/nbds/) or [ulib](https://code.google.com/p/ulib/)?

**Comment by yuyichao** [comment]
I didn't know such libraries exists.

1. `ulib`

    A breif look at the implementation doesn't seems like it is thread safe. (It's basically the same with the current symbol table plus using callback to compare the node and having to look up twice for insertion)

2. `nbds`

    The project claims itself to be pre-alpha. It also claim to be x86/x64 only and I'm pretty sure this PR doesn't have much problem there.

3. `libcuckoo`

    The biggest problem is that it's a c++ template. We also might need to allocate twice as much memory since it also store the keys.

**Comment by vtjnash** [comment]
i think the travis failure was some sort of transient CI issue (exact message was "ERROR: could not open file sysimg.jl" from julia-debug, perhaps caused by julia reading the file at the same time?)

lgtm

**Comment by tkelman** [comment]
> perhaps caused by julia reading the file at the same time

That's as good a theory as any. We're effectively doing `make -j3 install` now, so the release and debug bootstraps are overlapping. Can anyone get this to happen locally? Might need a makefile tweak.

**Comment by vtjnash** [comment]
if true, it seems that would imply a docker bug, but that also doesn't seem terribly likely

**Comment by tkelman** [comment]
union filesystems can be a little strange, so I wouldn't rule it out.

**Comment by ArchRobison** [comment]
Commit looked plausible to me.  Eventually we should migrate to using C11 atomics so we're not hand-waving with volatile.

**Comment by yuyichao** [comment]
Agree, something with well defined memory order that is designed for multi-threading would be much better. For now I think `volatile` should do the right thing =).

**Pull request merged**